### PR TITLE
DO NOT MERGE YET :: bump up to newer versions of odo, oc, and...

### PR DESCRIPTION
**Pull request by nickboldt**
bump up to newer versions of odo, oc, and kubectl; reorder Dockerfile to save steps

Change-Id: I40c45127dc8894dc7b2d2bae4fe6479ecfe99646
Signed-off-by: nickboldt <nboldt@redhat.com>

### CODE REVIEW

**Comment by azatsarynnyy** [comment]
@nickboldt the build is failing as no curl in alpine.
I'm adding plugins team to reviewers.

**Comment by nickboldt** [comment]
Since the kubernetes connector is based on Fedora/buildah image, maybe we should move this one over to that same base too? Why have alpine here and Fedora there, since they're almost the same build process for both images?

File: Dockerfile
Comment (benoitf) [inline Dockerfile:22]: Hello, is there a reason to move it to an upper level ? (it reduces caching layer) 
I suppose it's to merge ENV instruction into one?
and change ENV format ? (it's a lint error ? )

File: Dockerfile
Comment (benoitf) [inline Dockerfile:35]: do we need this change ?

File: Dockerfile
Comment (benoitf) [reply Dockerfile:22]: I mean, if each committer is applying his preferences on ENV params and cli options each time there is a bump in versions it might become a lot of changes for nothing (I have no preferences at all, just wanting the smallest changes each time)

**Comment by azatsarynnyy** [comment]
> Why have alpine here and Fedora there

@nickboldt at the beginning, both were based on alpine. Then, we needed to install Buildah in k8s sidecar and switched it to Fedora as there wasn't a Buildah package for Alpine at that time.

File: Dockerfile
Comment (nickboldt) [reply Dockerfile:35]: that depends. do you NEED CVE fixes? I mean there's a reason why oc 3.11 is on its 170th z-stream bugfix release. Is there a reason why the upstream vsix team is still using 3.11.0 instead of 3.11.170? Are they afraid of bugfixes? :D

File: Dockerfile
Comment (nickboldt) [reply Dockerfile:22]: I was trying to collapse steps. I thought you'd approve of having all the ENVs declared in a single place instead of in several commands. Why define HOME and then do some work, and then define more ENVs later? It's just... smelly code.

Comment (nickboldt) [reply Dockerfile:22]: BTW if you want smaller changes, we should be working in master branch and branching from that for each new release, rather than creating an empty branch and adding new files. 

Compare:
https://github.com/che-dockerfiles/che-sidecar-openshift-connector/commits/0.1.1/Dockerfile (one commit)
https://github.com/che-dockerfiles/che-sidecar-openshift-connector/commits/0.1.2/Dockerfile (one commit)
(no git blame to diff versions and see why lines have changed)

vs...

https://github.com/eclipse/che-machine-exec/commits/7.7.0/Dockerfile (a smaller series of commits)
https://github.com/eclipse/che-machine-exec/commits/7.8.0/Dockerfile (a few more commits, with the ability to do git blame)

This "all files are new" approach means it's super difficult to know what's changed between releases. Why are we doing it that way? For obfuscation?

**Comment by nickboldt** [comment]
So... why not move both to buildah? Dogfooding, remember? :D

File: Dockerfile
Comment (benoitf) [reply Dockerfile:35]: I mean why do we need changes if we bump OC_VERSION only ( not URL)

File: Dockerfile
Comment (benoitf) [reply Dockerfile:22]: I was referencing ENV with quotes vs ENV with spaces

Comment (benoitf) [reply Dockerfile:22]: You may also look at https://github.com/eclipse/che/issues/15949

File: Dockerfile
Comment (nickboldt) [reply Dockerfile:35]: OC 3.11.0 can be downloaded from here: https://github.com/openshift/origin/releases/download/${OC_VERSION}/openshift-origin-client-tools-${OC_VERSION}-${OC_TAG}-linux-64bit.tar.gz

But all the other 3.11.x versions are NOT in github releases. Instead they are here:

https://mirror.openshift.com/pub/openshift-v3/clients/${OC_VERSION}/linux/oc.tar.gz

They are also packaged differently, which is why the untar command is different.

File: Dockerfile
Comment (nickboldt) [reply Dockerfile:22]: Quotes are optional, but recommended (shellcheck, probably). 

But if you use spaces you can only do one ENV declaration per line. Using equal signs allow you to chain them into a single layer.

**Review by benoitf — APPROVED** [review]

**Comment by nickboldt** [comment]
Before merging this we should make sure the vsix extension doesn't try to download alternative versions of binaries if it can't find the One True Version hardcoded into its spec. Mykhailo was doing some investigation today into this problem and we night need new releases of upstream vsix before we can bump versions of oc/odo/kubectl/helm binaries in the sidecars. cc: @mkuznyetsov

File: Dockerfile
Comment (ericwill) [inline Dockerfile:32]: Shouldn't we also use `--no-cache` here?

File: Dockerfile
Comment (ericwill) [inline Dockerfile:11]: If we're updating all the deps, may as well update the base image to the latest (I think it's 3.11.3 these days).

**Review by ericwill — COMMENTED** [review]
I investigated the dependency downloading issue -- yes the extension does try to download its own version if it does not find the one true version. This is reflected by an error message in Che if you try to run the extension with this updated Dockerfile.

The good news is that [upstream the extension has adopted a range specifier](https://github.com/redhat-developer/vscode-openshift-tools/commit/b6af239c19c417fad04f1abf9a9f69b15f59fd32) for deps like odo, so we shouldn't have this problem going forward. The caveat of course is that we have to upgrade to the latest version of the extension, which means waiting until 0.1.4 is released. Maybe it's best we just do that instead of trying to update this one? Either that or we omit bumping the odo dep for this change, as that seems to be one breaking things.

File: Dockerfile
Comment (ericwill) [reply Dockerfile:32]: Of course the GitHub UI has bested me once again. This comment was intended for the `apk --update --allow-untrusted add glibc-${GLIBC_VERSION}.apk` line.

File: Dockerfile
Comment (mkuznyetsov) [inline Dockerfile:15]: since this image is for openshift-connector plugin 0.1.2 version, then ODO should be 1.0.1 version, since plugin is configured to work with that one

**Pull request closed without merging**
